src/msg_recorder/recorder.h

```diff
@@ -18,39 +24,61 @@ class MessageRecorder : public CRNamedNode<MessageRecorder> {
    public:
     using Base = CRNamedNode<MessageRecorder>;
 
-    explicit MessageRecorder(const std::filesystem::path& record_dir_prefix, std::shared_ptr<JobRunner> runner);
+    explicit MessageRecorder(const RecorderConfig& recorder_config, std::shared_ptr<JobRunner> runner);
 
     MessageRecorder(const MessageRecorder&) = delete;
-    MessageRecorder(MessageRecorder&&)      = default;
+    MessageRecorder(MessageRecorder&&)      = delete;
     MessageRecorder& operator=(const MessageRecorder&) = delete;
     MessageRecorder& operator=(MessageRecorder&&) = delete;
 
     ~MessageRecorder();
 
+    void StopMainLoop() override;
+
     template<CRMessageType message_t>
     void RegisterChannel(const channel_subid_t subid, const std::string& alias = "");
 
+    void MakeSnapshot();
+
     std::filesystem::path GetRecordDir() const;
 
    private:
     using msg_serializer = std::function<std::string(const CRMessageBasePtr&)>;
 
     RecordFile* CreateFile(const std::string& message_type, const channel_subid_t subid, const std::string& alias);
 
-    static std::string RecordDirNameGenerator();
+    void SetSnapshotInterval(const RecorderConfig& recorder_config);
+    void GenerateSnapshot(const std::size_t max);
 
-    const std::filesystem::path              record_dir_;
-    std::vector<std::unique_ptr<RecordFile>> files_;
+    void SnapshotWorkerStart();
+    void SnapshotWorkerEnd();
+
+    static std::string RecordDirNameGenerator();
+    static std::string SnapshotDirNameGenerator();
+
+    const std::filesystem::path                  record_dir_;
+    std::vector<std::unique_ptr<RecordFile>>     files_;
+    std::shared_ptr<cris::core::JobRunnerStrand> record_strand_;
+
+    const std::size_t                 keep_max_{48};
+    std::chrono::seconds              snapshot_interval_;
+    std::thread                       snapshot_thread_;
+    std::atomic<bool>                 snapshot_shutdown_flag_{false};
+    std::atomic<bool>                 snapshot_pause_flag_{false};
```

Comment:
wait for the `AddJobToRunner` lambda logic to be finished, which is `GenerateSnapshot`, since we pass `this` to the lambda and it's better to wait for the sake of safety, as we said before